Unbreak builds with old C++ compilers

andyross opened the pull request:
Cadence xcc can't handle some of the recent C++11 code.  (There are third party out of tree SOF modules using C++ too that aren't ready for xt-clang).  Let's slow walk the transition a bit and recompatibilize things for older compilers.

andyross committed:
cmake/compiler: Limit warning flag usage to compatible toolchains

New C++ versions have deprecated "register" variables and restricted
"volatile" semantics, so new gcc's will emit warnings when they see
that syntax.  Zephyr uses both in our C headers (though we should
probably get rid of register and unify with C++'s volatile model), so
we're disabling the resulting warnings.

But OLD gcc variants (like xcc, sigh) don't understand new -Wvolatile
and -Wregister on the command line, so they get confused.  Limit the
uses to the standard versions for which gcc would emit warnigns; xcc
doesn't support those anyway.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
minimal libcpp: Un-11-ify cstddef

Don't expose C++11 APIs unless we're building for that platform

Signed-off-by: Andy Ross <andyross@google.com>

squashme cdecl

andyross committed:
include/sys/util: Rework _Generic & C++11 usage

The recently-added LOG2 family of macros were implemented using either
a compile-time expanded C11 _Generic expression or a C++11 constexpr
template.  As it happens, Cadence xt-xcc supports neither of these
constructs, and integration for the next-generation xt-clang[1] hasn't
arrived yet.

Let's back out to a more portable (heh) implementation using gcc
__typeof__.  This keeps the advantage of expanding to a single
__builtin_clz() for common case usage and is result-compatible with
the new LOG2/LOG2CEIL/NHPOT.

The one downside is that the generic code was able to be used in a
constant context, something that statement expressions can't do.  So
the resulting macros are (like the earlier LOG2CEIL was) unhygienic
and multi-expand their arguments.  I added a warning to this effect in
the docs.

[1] Which is still based on an ancient clang tree, but more recent
than their GCC 4.2 compiler.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Back link to a bunch of discussion in #53405 which is relevant

cfriedt reviewed (dismissed):
Well done, @andyross - thanks!

cfriedt commented:
The one thing I would ask if you are willing to do another rev, would be to add a few test permutations to `tests/lib/cpp/cxx/testcase.yaml` specifying the STD_CPP98, STD_CPP11, STD_CPP14, STD_CPP98, ... that simply include all of the mission critical headers.
cc @stephanosio

andyross commented:
Just stumbled on #55204 with more relevant discussion

andyross committed:
tests/lib/cpp/cxx: Add compatibility cases for C++ standards

Add some cases to enumerate all the C++ standard variants that the SDK
gcc supports (there are MANY) to prevent compatibility regressions in
the OS headers.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
> add a few test permutations 

That seems like a great idea.  See final commit, unless I messed things up.

nordicjm commented:
GCC 4.2.0 is at this point, 16 years old. So my question is what problems or inefficiencies are being introduced by rolling back the standard level we are using and having to support an older version? And for how much longer do we need to support such an old compiler?

andyross commented:
Until in-tree projects aren't using it, basically.  I don't like xcc either, but it's what existing source code using current Zephyr is being built with.   If we don't want to support it we should remove the existing integration and tell SOF to fork the project or whatever.

And FWIW: in this case I think the justification was pretty thin for the features we were using.  The constexpr template wasn't providing much value, and the other stuff was just bitrot.

I do see that the new C++98 test variant pulled up more failures though, so I need to take a look at those.

andyross committed:
drivers/gna: No empty structs

Empty structs are incompatible when building with C++, which specifies
their size as 1 byte.  Technically they are illegal in C; GCC has
always supported them as an extension, but with a size of *zero*
bytes.

The upshot is that we can't have them in headers that may be presented
to a C++ compiler.  Add a placeholder field.

We only get the resulting compiler warning with -std=c++98 it seems
like, which is why this went undetected.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
sys/cbprintf: Limit C++ rvalue references to compatible standards

This feature didn't appear until C++11, so don't expose our wrapper
for it in C++98.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Two more patches to clean up issues detected with CONFIG_STD_CPP98=y.

andyross committed:
toolchain/gcc: Limit use of _Static_assert

This C11 feature doesn't work when building C++ code, but the C++
version is based on C++11's static_assert(), which we don't have in
-std=c++98 either.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Ugh, one more.  Pretty sure this is it.

andyross committed:
doc: Clarify C++ standards support

We weren't explicit about what we mean by "C++ support" as it relates
to standards versions, and it's been causing some friction in the tree
vs. some of our cruftier toolchains.  Add a paragraph to clarify
things.

Fixes #55204

Signed-off-by: Andy Ross <andyross@google.com>

stephanosio commented on include/zephyr/sys/cbprintf_cxx.h:
```diff
@@ -249,10 +249,12 @@ struct z_cbprintf_cxx_remove_reference < T & > {
 	typedef T type;
 };
 
+#ifndef CONFIG_STD_CPP98
```
I would prefer to generalise this using the `__cplusplus` macro since it is more flexible and more or less what we have been using for checking the effective C++ standard in other places.

`__cplusplus >= 201103L`

andyross replied:
It's not uniformly supported in that way though, xcc defines it as just "1".  Which would actually work in this case (1 is less than 199711), but I doubt that's the only edge case.  What if a compiler supports C++17 but was passed -std=c++11, what do we get?  Do we know for sure?  Have we tested?

Given exactly the kind of nonsense we're dealing with here, I feel strongly that relying on our own abstraction is the right choice.

stephanosio replied:
> What if a compiler supports C++17 but was passed -std=c++11, what do we get?

`__cplusplus` should be defined as the effective standard version; so, that would be `201103L` if `-std=c++11`.

> Do we know for sure? Have we tested?

Yes, that is what the standard says and how it must be.

From ISO/IEC 14882:1998(E) 16.8 Predefined macro names:

> The name `__cplusplus` is defined to the value 199711L when compiling a C++
translation unit.

> It's not uniformly supported in that way though, xcc defines it as just "1". Which would actually work in this case (1 is less than 199711)

That is a compiler bug and, as you say, it works either way. You can expect any non-ancient compiler that supports C++11 and above to properly define this macro according to the ISO C++ standard.

andyross replied:
Changed, though this particular one gives me pretty bad ick.  Relying on compiler conformance for standards minutiae in areas that are **known** to be improperly implemented seems like the opposite of defensive/conservative best practice.

stephanosio reviewed (changes requested):
Looks ok overall; just some minor comments.

andyross commented on include/zephyr/toolchain/gcc.h:
```diff
@@ -70,11 +70,14 @@
  */
-#elif (__GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 6)) || \
-	(__STDC_VERSION__) >= 201100
+#elif !defined(CONFIG_STD_CPP98) && \
```
This bit I can change, because you're right that the C++ magic already happened.

andyross commented on tests/lib/cpp/cxx/testcase.yaml:
```diff
@@ -25,3 +25,33 @@ tests:
+    build_only: true
+    extra_configs:
+      - CONFIG_STD_CPP98=y
+  cpp.main.cpp11:
+    build_only: true
+    extra_configs:
+      - CONFIG_STD_CPP11=y
```
Only until the default changes, though.  If it does, and I apply this, then "cpp.main.cpp11" will be testing some standard other than C++11!  I think that's clearly not what we want; it's not illegal to specify a kconfig in two places.

stephanosio replied:
What I meant was to drop `cpp.main.cpp11` altogether. We do not need an additional testcase that tests the same thing that other testcases above test.

Maybe it would be helpful to add a comment saying "C++11 case is not explicitly tested because it is the default option and already tested by the above testcases."

andyross replied:
Fair enough

andyross commented on include/zephyr/sys/util.h:
```diff
@@ -592,34 +511,49 @@ char *utf8_trunc(char *utf8_str);
  */
 char *utf8_lcpy(char *dst, const char *src, size_t n);
 
+#define __z_log2d(x) (32 - __builtin_clz(x) - 1)
+#define __z_log2q(x) (64 - __builtin_clzll(x) - 1)
+#define __z_log2(x) (sizeof(__typeof__(x)) > 4 ? __z_log2q(x) : __z_log2d(x))
```
It's important to note here that for the actual features at hand, the GCC implementation is significantly **more** portable (identical six-line implementation for AFAIK literally every compiler ever used to build zephyr) than the first cut (50 lines containing two different implementations across variant generic APIs that broke at least one in-tree Zephyr user).

I mean... yes, standards are good.  But we need to be realistic about things too.

stephanosio replied:
> the GCC implementation is significantly **more** portable

Until you use something other than GCC (or a compiler that intentionally follows what GCC does; notoriously, Clang). It really is not "portable" unless you live in a world where there are only two compilers: GCC and Clang.

I have no problem with using the GNU extensions where it makes sense; but, we should always provide a standard conforming solution alongside.

stephanosio replied:
> I'm just telling you straight up that the definition for "portable" you have chosen to defend means "works on fewer compilers" and not "works on more compilers", which is IMHO exactly backwards.

Once again, the problem I have here is not the use of the `__typeof__` GNU extension -- it is the removal of the existing implementations that do not require `__typeof__`.

andyross commented on lib/cpp/minimal/include/cstddef:
```diff
@@ -15,11 +15,21 @@
+	typedef void *nullptr_t; /* no decltype() yet */
+#else
 	using nullptr_t = decltype(nullptr);
+#endif
```
Fixed

andyross commented:
Yet more stuff turned up that works on the SDK (and my local gcc 12.2.1) but not the CI host compiler.  May just have to keep banging on this until they all fall out.

cfriedt reviewed (approved):
LGTM. I think let's put out the current fire and then make iterative changes to move things to toolchains, if needed.

For line 516 though, wouldn't `sizeof(x)` suffice?

andyross commented:
sizeof() only works on lvalues or type names.  You can "sizeof(int)" or "sizeof(*foo)", but you can't "sizeof(*foo+2)" or "sizeof(atan(dy, dx))"

marc-hb commented:
It's often hard to get things "done" while getting them "right", especially in this area.

I'm impressed by how this PR does both - with a negative diff on top.

Thanks for the free class.
